naming/program/src/processor.rs

```diff
@@ -0,0 +1,217 @@
+use crate::{
+    instruction::NameRegistryInstruction,
+    state::get_seeds_and_key,
+    state::{write_data, NameRecordHeader, NameStatus},
+};
+use borsh::BorshDeserialize;
+use solana_program::{
+    account_info::{next_account_info, AccountInfo},
+    entrypoint::ProgramResult,
+    msg,
+    program::invoke_signed,
+    program_error::ProgramError,
+    program_pack::Pack,
+    pubkey::Pubkey,
+    system_instruction::create_account,
+};
+
+pub struct Processor {}
+
+impl Processor {
+    pub fn process_create(
+        program_id: &Pubkey,
+        accounts: &[AccountInfo],
+        hashed_name: Vec<u8>,
+        lamports: u64,
+        space: u32,
+    ) -> ProgramResult {
+        let accounts_iter = &mut accounts.iter();
+
+        let system_program = next_account_info(accounts_iter)?;
+        let payer_account = next_account_info(accounts_iter)?;
+        let name_account = next_account_info(accounts_iter)?;
+        let name_owner = next_account_info(accounts_iter)?;
+        let name_class = next_account_info(accounts_iter)?;
+        let parent_name_account = next_account_info(accounts_iter)?;
+        let parent_name_owner = next_account_info(accounts_iter).ok();
+
+        let (name_account_key, seeds) = get_seeds_and_key(
+            &program_id,
+            hashed_name,
+            Some(name_class.key),
+            Some(parent_name_account.key),
+        );
+
+        // Verifications
+        if name_account_key != *name_account.key {
+            msg!("The given name account is incorrect.");
+            return Err(ProgramError::InvalidArgument);
+        }
+        if name_account.data.borrow().len() > 0
+            && name_account.data.borrow()[0] != (NameStatus::Uninitialized as u8)
+        {
+            msg!("The given name account already exists.");
+            return Err(ProgramError::InvalidArgument);
+        }
+        if *name_class.key != Pubkey::default() && !name_class.is_signer {
+            msg!("The given name class is not a signer.");
+            return Err(ProgramError::InvalidArgument);
+        }
+        if *parent_name_account.key != Pubkey::default() && !parent_name_owner.unwrap().is_signer {
+            msg!("The given parent name account owner is not a signer.");
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        if name_account.data.borrow().len() == 0 {
+            // Issue the name registry account
+            let create_account_instruction = create_account(
```

Comment:
So this is annoying but instead of `system_instruction::create_account`, we should instead issue these three in instructions:
* system_instruction::transfer
* system_instruction::allocate
* system_instruction::assign

This is exactly what `create_account` does internally, *except* that `create_account` will fail if a `system_instruction::transfer` has previously been used to send lamports to the account.  This means that I can prevent you from grabbing a name simply by running `solana transfer`

Example at https://github.com/solana-labs/solana-program-library/blob/0f4f2b8de909f1bf356260bde8b401813c3dfdec/associated-token-account/program/src/processor.rs#L58-L98

Reply:
Oh I didn't know about this limitation of `create`, thanks!